### PR TITLE
add experiment tags to MLFlow storage backend

### PR DESCRIPTION
**Pull request by ankit-db**
## What changes are proposed in this pull request?
 
This is a PR for adding experiment tags to MLFlow storage. See #1632 for the sibling PR on the client side. After discussion with Matei, we're going to go with a separated API, meaning we will have set_experiment_tag separate from the set_tag semantics for the run tags. This PR adds tests so that this workflow is well tested, along with a SQLAlchemy migration that will migrate existing databases to add the new experiment_tags table. 

## How is this patch tested?
 
Unit testing and manual testing
 
### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
We are adding the ability to add tags to your experiments in MLFlow! This will mean that you can add notes and other key-value tags to your experiments and see them via the API. 
 
### What component(s) does this PR affect?
 
- [x] UI
- [ ] CLI 
- [x] API 
- [x] REST-API 
- [ ] Examples 
- [x] Docs
- [ ] Tracking
- [x] Projects 
- [ ] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [x] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by ankitmathur-db** [comment]
New PR because Github was being super weird about the old one.

File: mlflow/store/file_store.py [hunk 1]
```diff
@@ -575,6 +598,19 @@ def log_param(self, run_id, param):
...
 
+    def set_experiment_tag(self, experiment_id, tag):
+        """
+        Set a tag for the specified experiment
+
+        :param experiment_id: String ID of the experiment
+        :param tag: ExperimentRunTag instance to log
+        """
+        _validate_tag_name(tag.key)
+        # experiment validation happens in get_experiment_tag_path
```
Comment (dbczumar) [inline mlflow/store/file_store.py]: We aren't validating that the experiment is active. The following code sets a tag on a deleted experiment successfully:
```
import mlflow
from mlflow.tracking.client import MlflowClient

mlflow.set_experiment("DogExperiment")
client = MlflowClient()
exp = client.get_experiment("1")
print(exp)

client.delete_experiment("1")

client.set_experiment_tag("1", "DOG", "CAT")

exp = client.get_experiment("1")
print(exp)
```

Log output:
```
$ python filesetinactive.py
INFO: 'DogExperiment' does not exist. Creating a new experiment
<Experiment: artifact_location='file:///Users/czumar/mlflow/DEBUG/exptag/mlruns/1', experiment_id='1', lifecycle_stage='active', name='DogExperiment', tags={}>
<Experiment: artifact_location='file:///Users/czumar/mlflow/DEBUG/exptag/mlruns/1', experiment_id='1', lifecycle_stage='deleted', name='DogExperiment', tags={'DOG': u'CAT'}>
```

We should fix this and add a test to ensure that an exception is thrown instead.

File: mlflow/store/dbmodels/models.py [hunk 2]
```diff
@@ -188,6 +189,47 @@ def to_mlflow_entity(self):
...
+    DB model for :py:class:`mlflow.entities.RunTag`.
+    These are recorded in ``experiment_tags`` table.
+    """
+    __tablename__ = 'experiment_tags'
+
+    key = Column(String(250))
+    """
+    Tag key: `String` (limit 250 characters). *Primary Key* for ``tags`` table.
+    """
+    value = Column(String(250), nullable=True)
```
Comment (dbczumar) [inline mlflow/store/dbmodels/models.py]: The maximum tag length configured here is less than the maximum length in the R documentation you included in https://github.com/mlflow/mlflow/pull/1632. I realize that the `value` column semantics for `SqlTag` also violate the documentation. IMO, we should use a maximum size of at least 500 bytes.

Comment (dbczumar) [inline mlflow/store/dbmodels/models.py]: As a test, I attempted to log an experiment tag with a 500-byte value using the current configuration against MySQL. Here's the code I ran:
```
import mlflow
from mlflow.tracking.client import MlflowClient

mlflow.set_tracking_uri("mysql://root:password@localhost:33060/exptags")

client = MlflowClient()

mlflow.set_experiment("Dogface")
exp = client.get_experiment("1")
print(exp)

long_value = "a" * 500
print(len(long_value))

client.set_experiment_tag(
    "1",
    "key",
    long_value)

longtag = client.get_experiment("1").tags["key"]
print(len(longtag))
```

Interestingly, when running this code against a fresh database seemed to truncate the tag; here was the STDOUT:
```
$ python sqllong.py
2019/07/19 14:24:48 INFO mlflow.store.sqlalchemy_store: Creating initial MLflow database tables...
2019/07/19 14:24:48 INFO mlflow.store.db.utils: Updating database tables at mysql://root:password@localhost:33060/exptags
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade  -> 451aebb31d03, add metric step
INFO  [alembic.runtime.migration] Running upgrade 451aebb31d03 -> 90e64c465722, migrate user column to tags
INFO  [alembic.runtime.migration] Running upgrade 90e64c465722 -> 181f10493468, allow nulls for metric values
INFO  [alembic.runtime.migration] Running upgrade 181f10493468 -> df50e92ffc5e, Add Experiment Tags Table
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO: 'Dogface' does not exist. Creating a new experiment
<Experiment: artifact_location='./mlruns/1', experiment_id='1', lifecycle_stage='active', name='Dogface', tags={}>
500
250
```

However, attempting to run the exact same code again yields the following error output:
```
 $ python sqllong.py
<Experiment: artifact_location='./mlruns/1', experiment_id='1', lifecycle_stage='active', name='Dogface', tags={'key': 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'}>
500
Traceback (most recent call last):
  File "sqllong.py", line 21, in <module>
    long_value)
  File "/Users/czumar/mlflow/mlflow/tracking/client.py", line 187, in set_experiment_tag
    self.store.set_experiment_tag(exp_id, tag)
  File "/Users/czumar/mlflow/mlflow/store/sqlalchemy_store.py", line 492, in set_experiment_tag
    value=tag.value))
  File "/Users/czumar/anaconda2/lib/python2.7/contextlib.py", line 24, in __exit__
    self.gen.next()
  File "/Users/czumar/mlflow/mlflow/store/sqlalchemy_store.py", line 156, in make_managed_session
    raise MlflowException(message=e, error_code=INTERNAL_ERROR)
mlflow.exceptions.MlflowException: (MySQLdb._exceptions.DataError) (1406, "Data too long for column 'value' at row 1")
[SQL: UPDATE experiment_tags SET value=%s, experiment_id=%s WHERE experiment_tags.`key` = %s AND experiment_tags.experiment_id = %s]
[parameters: ('aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa ... (202 characters truncated) ... aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa', '1', 'key', 1)]
(Background on this error at: http://sqlalche.me/e/9h9h)
```

We should figure out what's causing this behavioral inconsistency and fix it. Can you also try to reproduce the problem with run tags?

File: tests/resources/db/latest_schema.sql [hunk 3]
```diff
@@ -32,8 +41,8 @@ CREATE TABLE runs (
 	experiment_id INTEGER, 
 	CONSTRAINT run_pk PRIMARY KEY (run_uuid), 
 	FOREIGN KEY(experiment_id) REFERENCES experiments (experiment_id), 
-	CONSTRAINT status CHECK (status IN ('SCHEDULED', 'FAILED', 'FINISHED', 'RUNNING')), 
 	CONSTRAINT source_type CHECK (source_type IN ('NOTEBOOK', 'JOB', 'LOCAL', 'UNKNOWN', 'PROJECT')), 
+	CONSTRAINT status CHECK (status IN ('SCHEDULED', 'FAILED', 'FINISHED', 'RUNNING')),
```
Comment (dbczumar) [inline tests/resources/db/latest_schema.sql]: Nit: Why was this line moved?

File: tests/resources/db/latest_schema.sql [hunk 4]
```diff
@@ -16,6 +16,15 @@ CREATE TABLE experiments (
 )
 
 
+CREATE TABLE experiment_tags (
+	key VARCHAR(250) NOT NULL, 
+	value VARCHAR(250),
```
Comment (dbczumar) [inline tests/resources/db/latest_schema.sql]: If we end up using a larger maximum tag length limit, we should update `latest_schema` accordingly.

**Review by dbczumar — CHANGES_REQUESTED** [review]
@ankitmathur-db Nice job! Left a few comments

File: mlflow/store/file_store.py [hunk 5]
```diff
@@ -410,19 +419,28 @@ def _get_run_files(self, run_uuid, resource_type):
...
                                   databricks_pb2.INVALID_STATE)
+
+        _, run_dir = self._find_run_root(run_uuid)
+        # run_dir exists since run validity has been confirmed above.
+        return self._get_metric_files(run_dir, resource_type)
+
+    def _get_experiment_files(self, experiment_id):
+        _validate_experiment_id(experiment_id)
+        experiment_dir = self._get_experiment_path(experiment_id, assert_exists=True)
+        return self._get_metric_files(experiment_dir, "tag")
```
[the remark this answers is not in the document]
Comment (ankitmathur-db) [reply mlflow/store/file_store.py]: The only non-generic portion of the resource function is the subfolder_name - I think it doesn't make too much sense to copy the subsequent logic to both functions separately just because it happens to be for runs as well as experiments. I did refactor some of the resource type logic out, so hopefully that should resolve the concerns about logical isolation.

File: tests/resources/db/latest_schema.sql (hunk 3 above)
Comment (ankitmathur-db) [reply tests/resources/db/latest_schema.sql]: Autogen

File: mlflow/store/file_store.py [hunk 6]
```diff
@@ -410,19 +419,28 @@ def _get_run_files(self, run_uuid, resource_type):
...
+
+    def _get_metric_files(self, root_dir, resource_type):
```
[the remark this answers is not in the document]
Comment (ankitmathur-db) [reply mlflow/store/file_store.py]: See discussion below - refactored, but in a way that reuses a nice portion of the code.

File: mlflow/store/file_store.py [hunk 7]
```diff
@@ -7,7 +7,7 @@
 import six
```
[the remark this answers is not in the document]
Comment (ankitmathur-db) [reply mlflow/store/file_store.py]: It is not experimental - the meta.yaml that's written just calls `dict` on the Experiment object, which, naturally, does not have any tags in it at the time of creation. At read time, we just poll for the new tags that were written and we override that parameter. I went ahead and just added code to manually delete this from the written dictionary, but I don't think this should cause any bugs otherwise.

File: mlflow/store/file_store.py (hunk 1 above)
Comment (ankitmathur-db) [reply mlflow/store/file_store.py]: Yeah I think I got lost in the validation logic - my bad! Fixed and added a test.

File: tests/resources/db/latest_schema.sql (hunk 4 above)
Comment (ankitmathur-db) [reply tests/resources/db/latest_schema.sql]: yup - this is autogen'd so, i will re-run that code

File: mlflow/store/dbmodels/models.py (hunk 2 above)
Comment (ankitmathur-db) [reply mlflow/store/dbmodels/models.py]: Updating experimentTag to be 500 - I think the runTag update should be in a separate PR.

Comment (ankitmathur-db) [reply mlflow/store/dbmodels/models.py]: Per offline discussion, this seems related to strict mode execution on the SQL queries. Solution is to do tag validation before setting.

File: mlflow/store/sqlalchemy_store.py [hunk 8]
```diff
@@ -466,7 +475,28 @@ def log_param(self, run_id, param):
...
+        Set a tag for the specified experiment
+
+        :param experiment_id: String ID of the experiment
+        :param tag: ExperimentRunTag instance to log
+        """
+        with self.ManagedSessionMaker() as session:
+            experiment = self._get_experiment(session,
+                                              experiment_id,
+                                              ViewType.ALL).to_mlflow_entity()
+            self._check_experiment_is_active(experiment)
```
[the remark this answers is not in the document]
Comment (ankitmathur-db) [reply mlflow/store/sqlalchemy_store.py]: Yup - done!

File: mlflow/protos/service.proto [hunk 9]
```diff
@@ -776,6 +806,22 @@ message LogParam {
...
+message SetExperimentTag {
+  option (scalapb.message).extends = "com.databricks.rpc.RPC[$this.Response]";
+
+  // ID of the experiment under which to log the tag. Must be provided.
+  optional string experiment_id = 1 [(validate_required) = true];
+
+  // Name of the tag. Maximum size is 255 bytes.
+  optional string key = 2 [(validate_required) = true];
+
+  // String value of the tag being logged. Maximum size is 5000 bytes.
```
Comment (dbczumar) [inline mlflow/protos/service.proto]: Is the maximum size really 5000 bytes?

File: mlflow/protos/service.proto [hunk 10]
```diff
@@ -776,6 +806,22 @@ message LogParam {
   }
 }
 
+message SetExperimentTag {
+  option (scalapb.message).extends = "com.databricks.rpc.RPC[$this.Response]";
+
+  // ID of the experiment under which to log the tag. Must be provided.
+  optional string experiment_id = 1 [(validate_required) = true];
+
+  // Name of the tag. Maximum size is 255 bytes.
```
Comment (dbczumar) [inline mlflow/protos/service.proto]: 255 bytes or 250 bytes?

File: mlflow/store/sqlalchemy_store.py [hunk 11]
```diff
@@ -366,6 +365,13 @@ def _check_run_is_active(self, run):
                                   .format(run.run_uuid, run.lifecycle_stage),
                                   INVALID_PARAMETER_VALUE)
 
+    def _check_experiment_is_active(self, experiment):
+        if experiment.lifecycle_stage != LifecycleStage.ACTIVE:
+            raise MlflowException("The experiment {} must be in 'active' state. "
```
[the remark this answers is not in the document]
Comment (ankitmathur-db) [reply mlflow/store/sqlalchemy_store.py]: Sure, I changed them for all the other functions too then

File: mlflow/entities/experiment_tag.py [hunk 12]
```diff
@@ -0,0 +1,34 @@
+from mlflow.entities._mlflow_object import _MLflowObject
+from mlflow.protos.service_pb2 import ExperimentTag as ProtoExperimentTag
+
+
+class ExperimentTag(_MLflowObject):
+    """Tag object associated with a run."""
```
Comment (dbczumar) [inline mlflow/entities/experiment_tag.py]: `run` >> `experiment`

File: mlflow/protos/service.proto [hunk 13]
```diff
@@ -246,6 +246,25 @@ service MlflowService {
     };
   }
 
+  // Set a tag on an experiment. Tags are run metadata that can be updated on an experiment.
```
Comment (dbczumar) [inline mlflow/protos/service.proto]: Let's change the second sentence to read `Experiment tags are metadata that can be updated.`.

File: mlflow/utils/validation.py [hunk 14]
```diff
@@ -98,6 +100,15 @@ def _validate_tag(key, value):
     _validate_length_limit("Tag value", MAX_TAG_VAL_LENGTH, value)
 
 
+def _validate_experiment_tag(key, value):
+    """
+    Check that a tag with the specified key & value is valid and raise an exception if it isn't.
+    """
+    _validate_tag_name(key)
+    _validate_length_limit("Tag key", MAX_EXPERIMENT_TAG_KEY_LENGTH, key)
```
Comment (dbczumar) [inline mlflow/utils/validation.py]: While we're at it, can we update the error message for `_validate_length_limit` (https://github.com/mlflow/mlflow/blob/067eea7ebb3bf9e0b0c164958eb0b619e3bf5119/mlflow/utils/validation.py#L122) to truncate the value emitted in the error message? Attempting to set a tag with a value of length 50,000 prints 50,000 characters to the console.

**Review by dbczumar — APPROVED** [review]
LGTM with nits. Nice job @ankitmathur-db!

File: mlflow/utils/validation.py (hunk 14 above)
Comment (dbczumar) [reply mlflow/utils/validation.py]: Displaying the first 250 characters is probably sufficient

**Review by dbczumar — CHANGES_REQUESTED** [review]
Did a bit of further testing and found an issue with the FileStore when attempting to use the SearchRuns API on an experiment with tags. Looks like SearchRuns (and `MlflowClient.list_run_infos`) attempts to parse the `tags` directory as a run:

```
In [11]: client.list_run_infos("1")
WARNING:root:Malformed run 'tags'. Detailed error Yaml file '/tmp/mlruns/1/tags/meta.yaml' does not exist.
Traceback (most recent call last):
  File "/Users/czumar/mlflow/mlflow/store/file_store.py", line 561, in _list_run_infos
    run_info = self._get_run_info(r_id)
  File "/Users/czumar/mlflow/mlflow/store/file_store.py", line 412, in _get_run_info
    meta = read_yaml(run_dir, FileStore.META_DATA_FILE_NAME)
  File "/Users/czumar/mlflow/mlflow/utils/file_utils.py", line 159, in read_yaml
    raise MissingConfigException("Yaml file '%s' does not exist." % file_path)
MissingConfigException: Yaml file '/tmp/mlruns/1/tags/meta.yaml' does not exist.
Out[11]: [<RunInfo: artifact_uri='file:///tmp/mlruns/1/7228b542e76c4eeab5dcbbda97e03805/artifacts', end_time=1564117553557, experiment_id='1', lifecycle_stage='active', run_id='7228b542e76c4eeab5dcbbda97e03805', run_uuid='7228b542e76c4eeab5dcbbda97e03805', start_time=1564117553375, status='FINISHED', user_id='czumar'>]
```

**Comment by ankitmathur-db** [comment]
Thanks for the catch @dbczumar - repro'd and fixed. Could you take a look?

**Comment by ankitmathur-db** [comment]
Test is at: https://github.com/mlflow/mlflow/pull/1632/commits/da69a81f600b52321cf313c3b9f542c8ee61984c

**Review by dbczumar — APPROVED** [review]
LGTM!
